src/dart/lib/decoder.dart

```diff
@@ -0,0 +1,894 @@
+      ]),
+      List<Color>.generate(_IconVGMachine.kCREGLength, (int index) => Color(0xFF000000 + (2 * index + 1))),
+    );
+    assert(testMachine.decodeColor1() == 0x000000FF);
+    assert(testMachine.decodeColor1() == 0x00FFFFFF);
+    assert(testMachine.decodeColor1() == 0x40FFC0FF);
+    assert(testMachine.decodeColor1() == 0xC08040FF);
+    assert(testMachine.decodeColor1() == 0xFFFFFFFF);
+    assert(testMachine.decodeColor1() == 0xC0C0C0C0);
+    assert(testMachine.decodeColor1() == 0x80808080);
+    assert(testMachine.decodeColor1() == 0x00000000);
+    assert(testMachine.decodeColor1() == 0x000001FF);
+    assert(testMachine.decodeColor1() == 0x000003FF);
+    assert(testMachine.decodeColor1() == 0x000005FF);
+    assert(testMachine.decodeColor1() == 0x00007FFF);
+    assert(testMachine.decodeColor1() == 0x000001FF);
+    assert(testMachine.decodeColor1() == 0x000003FF);
+    assert(testMachine.decodeColor1() == 0x000005FF);
+    assert(testMachine.decodeColor1() == 0x00007FFF);
+    assert(testMachine.decodeColor2() == 0x338800FF);
+    assert(testMachine.decodeDirectColor3() == 0x306607FF);
+    assert(testMachine.decodeColor4() == 0x30660780);
+    testMachine.customPalette[2] = 0xFF9000FF; // "opaque orange"
+    assert(testMachine.decodeIndirectColor3() == 0x40240040); // "25% opaque orange" (premultiplied)
+    assert(testMachine.decodeNaturalNumber() == 0x14);
+    assert(testMachine.decodeNaturalNumber() == 0x20D6);
+    assert(testMachine.decodeNaturalNumber() == 0xFE00001);
+    assert(testMachine.decodeRealNumber() == 20.0);
+    assert(testMachine.decodeRealNumber() == 8406.0);
+    assert(testMachine.decodeRealNumber() == 1.000000476837158203125);
+    assert(testMachine.decodeCoordinateNumber() == 7.0);
+    assert(testMachine.decodeCoordinateNumber() == 7.5);
+    assert(testMachine.decodeCoordinateNumber() == 7.5);
+    assert(testMachine.decodeZeroToOneNumber() == 15.0 / 360.0);
+    assert(testMachine.decodeZeroToOneNumber() == 40.0 / 360.0);
+    assert(testMachine.decodeZeroToOneNumber() == 0.00277777761220932); // approx 1.0/360.0
+  }
+}
+
+enum _RenderingMode { styling, drawing }
+enum _DrawingCommand { other, quadratic, cubic }
+
+class _IconVGMachine {
+  _IconVGMachine(this.bytes, {
+    List<Color>? palette,
+    required double height,
+    required this.clipBehavior,
+  }) : assert(palette == null || palette.length <= kCREGLength) {
+    checkSignature();
+    if (palette != null) {
+      applyCustomPalette(palette);
+    }
+    readMetadata();
+    if (foundPalette) {
+      assert(CREG.length == customPalette.length);
+      CREG.setRange(0, customPalette.length, customPalette);
+    }
+    if (!foundViewBox) {
+      minX = -32.0;
+      maxX = 32.0;
+      minY = -32.0;
+      maxY = 32.0;
+    }
+    viewBox = Rect.fromLTRB(minX, minY, maxX, maxY);
+    H = height.isFinite ? height : maxY - minY; // TODO(ianh): consider the conclusions in https://github.com/google/iconvg/issues/20
+    execute();
```

Comment:
this method probably shouldn't be public right? I understand this is a private class, but calling this a second time after construction will always be an error.

It might make more sense to just have a `computePicture` method or something like that instead of `execute`, and get rid of the late member field. I'd typically expect a ctor to be cheap, so that I can instantiate the machine wherever I want and control when/how often it actually does work. It may also be good to make that an async function, since you may at some point want it to do some of its work async in a separate isolate or something like that. Obviously can be avoided if you're super confident there will never be a good reason to have async parsing (for example, it has to interact with dart:ui objects too frequently to allow for any benefit from async computation/io).

Reply:
`execute` isn't really special here, the `checkSignature` and `readMetadata` methods are equally one-time-only. The whole class is really just a very large function with nested functions. I guess I could switch to that style instead, but I liked how the class syntax guarantees there's no stray imperative code after the constructor's `}` that might do something.